### PR TITLE
Refactors spell targeting to delegate it to a targeting datum

### PR DESCRIPTION
**Pull request by farie82**
## What Does This PR Do
Before everything else, it is a big PR and I would like a couple of TMs to flush out all the possible issues. I've tested quite a lot of the spells but not all extensively. I also fully understand if this PR stays open for a while as it touches quite some code.

It makes it so that all the spell targeting is done by the` datum/spell_targeting` variable.
This allows for easier changes to spells and makes spells such as vampire spells less snowflaky and easier to expand upon.

The core thought is that every spell type can define a `create_new_targeting` implementation which makes a targeting datum for that type of spell. Here it can pick a subtype of `datum/spell_targeting` and can customize the parameters. Every spell type only ever makes one of these objects which is then used by all other spells of that type. The `datum/spell_targeting` object is also stateless regarding the spell itself.

It also introduces `spell_handler`s which can delegate certain parts of spell code so it can be reused. It is for example used for vampire spells so that a jaunt (which reuses an existing spell) can still use the same vampire logic as all other spells

The blind spell is now also converted to be click targeted

Misc changes:
Removed dumbfire. Wasn't used and the code was horrible.
Fixed a runtime when using "Empathic Thought" on a mindless mob
Adds `pick_multiple_unique` to `lists.dm`
(ancient) Vampire enthrall now has a range of 1 instead of 7. No more tele bites
Removed explosion spell. It wasn't used.
Adds a unit test to see if all (non abstract) spells properly set their targeting
Updates vampire and morph spells

I might have missed some other functional changes. But I'll include those later when I find them.

## Why It's Good For The Game
Allows for easier customization of spells. Now you just have to make a spell that has an effect and you can reuse that spell for multiple targeting types. Want a remote EMP and also one centred on the user? You can. Just use `/datum/spell_targeting/self` and `/datum/spell_targeting/click` on a spell that makes an EMP on all targets (self or the click location).
It also makes the spell code quite a bit more readable in my opinion. It still ain't great but that would be a huge ass refactor to fix.

## Changelog
:cl:
add: A spells targeting is now delegated to a targeting datum. Allowing for easier customization of spells in the future.
del: The dumbfire spell is gone. It was as the name suggests, dumb.
del: Explosion spell which wasn't used
fix: Ancient vampire enthrall now uses 1 range distance instead of 7 while the text says he's biting somebody
tweak: The blind spell is now click based as well
/:cl:

### CODE REVIEW

File: code/datums/spell.dm [hunk 1]
```diff
@@ -191,17 +220,74 @@ GLOBAL_LIST_INIT(spells, typesof(/obj/effect/proc_holder/spell))
 	else
 		start_recharge()
 
+	if(!targeting_datums[type])
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spell.dm:223]: Should be fine since it is using assoc values instead of indexed values. Byond only throws the exceptions when you try to use a number in a list.

File: code/datums/spell_targeting/can_reach_around.dm [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/datum/spell_targeting/can_reach_around/choose_targets(mob/user, obj/effect/proc_holder/spell/spell, params, atom/clicked_atom)
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spell_targeting/can_reach_around.dm:1]: Yeah, I forgot to actually comment on it but I wasn't sure on this one either. It is only used for bats thus far.

File: code/datums/spells/wizard.dm [hunk 3]
```diff
@@ -65,7 +65,13 @@
 
 	sound = 'sound/items/bikehorn.ogg'
 
-/obj/effect/proc_holder/spell/targeted/inflict_handler/honk_missile
+/obj/effect/proc_holder/spell/projectile/honk_missile/create_new_targeting()
+	var/datum/spell_targeting/targeted/T = new()
+	T.allowed_type = /mob/living
+	T.max_targets = INFINITY
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spells/wizard.dm:71]: Yep yep. The old system required you to set `max_targets = 0`. But that seemed a bit odd to me. `INFINITY` is used now instead. I'll put a comment on the `max_targets` variable denoting this

File: code/datums/spells/wizard.dm [hunk 4]
```diff
@@ -272,12 +298,16 @@
...
-	range = 3
 
 	summon_type = list(/mob/living/simple_animal/hostile/creature)
 	cast_sound = 'sound/magic/summonitems_generic.ogg'
 
-/obj/effect/proc_holder/spell/targeted/trigger/blind
+/obj/effect/proc_holder/spell/aoe_turf/conjure/construct/creature/create_new_targeting()
+	var/datum/spell_targeting/aoe/turf/T = new()
+	T.range = 3
+	return T
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spells/wizard.dm:308]: Good catch! Seems I made a mistake copy-pasting the prev definition

File: code/datums/spell.dm [hunk 5]
```diff
@@ -149,18 +160,36 @@ GLOBAL_LIST_INIT(spells, typesof(/obj/effect/proc_holder/spell))
...
+/obj/effect/proc_holder/spell/proc/spend_spell_cost(mob/user)
+	switch(charge_type)
+		if("recharge")
+			charge_counter = 0 //doesn't start recharging until the targets selecting ends
+		if("charges")
+			charge_counter-- //returns the charge if the targets selecting fails
+		if("holdervar")
+			adjust_var(user, holder_var_type, holder_var_amount)
+	if(action)
+		action.UpdateButtonIcon()
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spell.dm:191]: Yep seems I don't need it there. Will change!

File: code/datums/spell_targeting/click.dm [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	if(max_targets <= length(found_others) + length(targets))
+		targets.Add(found_others)
+	else
+		switch(random_target_priority) //Add in the rest
+			if(SPELL_TARGET_RANDOM)
+				while(length(targets) < max_targets && length(found_others)) // Add the others
+					targets.Add(pick_n_take(found_others))
+			if(SPELL_TARGET_CLOSEST)
+				var/list/distances = list()
+				for(var/target in found_others) // maybe not needed TODO check
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spell_targeting/click.dm:31]: Forgot to confirm/deny that one. Will do today

File: code/datums/spells/bloodcrawl.dm [hunk 7]
```diff
@@ -3,24 +3,27 @@
...
+	T.range = 1
+	T.use_turf_of_user = TRUE
+	return T
+
+/obj/effect/proc_holder/spell/bloodcrawl/valid_target(obj/effect/decal/cleanable/target, user)
+	return target.can_bloodcrawl_in()
 
-/obj/effect/proc_holder/spell/bloodcrawl/perform(obj/effect/decal/cleanable/target, recharge = 1, mob/living/user = usr)
+/obj/effect/proc_holder/spell/bloodcrawl/perform(list/targets, recharge = 1, mob/living/user = usr)
+	var/obj/effect/decal/cleanable/target = targets[1]
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spells/bloodcrawl.dm:26]: Will always be a list with at least one target. I'll document this to make it more clear

File: code/datums/spells/emplosion.dm [hunk 8]
```diff
@@ -7,7 +7,11 @@
 
 	action_icon_state = "emp"
 
-/obj/effect/proc_holder/spell/targeted/emplosion/cast(list/targets, mob/user = usr)
+/obj/effect/proc_holder/spell/emplosion/create_new_targeting()
+	var/datum/spell_targeting/self/S = new()
+	return S
```
[the remark this answers is not in the document]
Comment (farie82) [reply code/datums/spells/emplosion.dm:12]: I'll do it for the `/datum/spell_targeting/self` types. But I'll leave it for the others who use the default values so it can be tweaked more easily later. It does make sense for the self type since it has no parameters

File: code/datums/spell_targeting/click.dm [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+	if(max_targets >= length(found_others) + length(targets))
+		targets.Add(found_others)
+	else
+		switch(random_target_priority) //Add in the rest
+			if(SPELL_TARGET_RANDOM)
+				while(length(targets) < max_targets && length(found_others)) // Add the others
+					targets.Add(pick_n_take(found_others))
+			if(SPELL_TARGET_CLOSEST)
+				// Take the first X. Byond's view/range procs already keep distance in mind. Will have a bias towards the left targets due to this
```
Comment (farie82) [inline code/datums/spell_targeting/click.dm:33]: The previous system also had this same bias. Just putting this here as a notice

File: code/modules/unit_tests/spell_targeting_test.dm [hunk 10]
```diff
@@ -0,0 +1,10 @@
+/datum/unit_test/spell_targeting/Run()
+	var/list/bad_spells = list()
+	for(var/obj/effect/proc_holder/spell/S as anything in typesof(/obj/effect/proc_holder/spell))
+		if(initial(S.name) == "Spell")
+			continue // Skip abstract spells
+		S = new S
+		if(!S.targeting)
+			bad_spells += S
+	if(length(bad_spells))
+		Fail("Spells without targeting found: [bad_spells.Join(",")]")
```
Comment (AffectedArc07) [inline code/modules/unit_tests/spell_targeting_test.dm:10]:
```suggestion
		Fail("Spells without targeting found: [bad_spells.Join(", ")]")
```

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]

**Review by AffectedArc07 — APPROVED** [review]
Approved the stuff I codeown.

This is not a merge-allowing review.

**Comment by farie82** [comment]
Big update for the PR due to recent merges. Both vampire and morph spells are now properly handled with the new system. I did need to introduce another addon for spells, `spell_handler`s, they are optional and allow developers to make modular spell logic datums. For example, the `spell_handler/vampire` allows the vampire version of jaunt to use the same logic as other vampire spells even though they are not the same type

File: code/datums/spell.dm [hunk 11]
```diff
@@ -131,18 +127,36 @@ GLOBAL_LIST_INIT(spells, typesof(/obj/effect/proc_holder/spell))
...
+	/// List with the handler datums per spell type. Key = src.type, value = the handler datum created by create_new_handler()
+	var/static/list/spell_handlers = list()
 
 /* Checks if the user can cast the spell
  * @param charge_check If the proc should do the cooldown check
  * @param start_recharge If the proc should set the cooldown
  * @param user The caster of the spell
 */
 /obj/effect/proc_holder/spell/proc/cast_check(charge_check = TRUE, start_recharge = TRUE, mob/user = usr) //checks if the spell can be cast based on its settings; skipcharge is used when an additional cast_check is called inside the spell
+	// SHOULD_NOT_OVERRIDE(TRUE) Todo for another refactor
```
Comment (farie82) [inline code/datums/spell.dm:159]: I couldn't dive into another rabbit hole anymore. This can be tackled later if needed. Only really needed to clean up the code. No functional changes

File: code/datums/spell_handler/vampire.dm [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+	vampire.bloodusable -= calculate_blood_cost(vampire)
+
+/datum/spell_handler/vampire/proc/calculate_blood_cost(datum/vampire/vampire)
+	var/blood_cost_modifier = 1 + vampire.nullified / 100
+	var/blood_cost = round(required_blood * blood_cost_modifier)
+	return blood_cost
+
+/datum/spell_handler/vampire/after_cast(list/targets, mob/user, obj/effect/proc_holder/spell/spell)
+	SSblackbox.record_feedback("tally", "vampire_powers_used", 1, "[spell]")
```
Comment (farie82) [inline code/datums/spell_handler/vampire.dm:48]: Found a bug here where it was not logged previously if it costs 0 blood

File: code/game/gamemodes/vampire/vampire_powers/hemomancer_powers.dm [hunk 13]
```diff
@@ -102,8 +99,14 @@
 	selection_activated_message = "<span class='notice'>You channel blood magics to weaken the bluespace veil. <B>Left-click to cast at a target area!</B></span>"
 	selection_deactivated_message = "<span class='notice'>Your magics subside.</span>"
 
+/obj/effect/proc_holder/spell/vampire/blood_tendrils/create_new_targeting()
+	var/datum/spell_targeting/click/T = new // TODO check
```
Comment (farie82) [inline code/game/gamemodes/vampire/vampire_powers/hemomancer_powers.dm:103]:
```suggestion
	var/datum/spell_targeting/click/T = new
```
Found another bug here. But that is unrelated to the PR

File: code/datums/spell_handler/vampire.dm (hunk 12 above)
Comment (farie82) [reply code/datums/spell_handler/vampire.dm]: Not a bug. Undone

**Comment by hal9000PR** [comment]
`All list input based vampire spells (such as disease touch)`  should be removed. as it no longer exists.

**Comment by farie82** [comment]
> `All list input based vampire spells (such as disease touch)` should be removed. as it no longer exists.

Good one. Completely forgot about removing that from the description

**Review by dearmochi — APPROVED** [review]
